utils/treesitter/src/scanner.c

```diff
@@ -54,34 +82,98 @@ static bool is_whitespace(char c) { return c == ' ' || c == '\n'; }
 // > fallbacks to the internal lexing mechanism.
 bool tree_sitter_carbon_external_scanner_scan(
     __attribute__((unused)) void* payload,
-    __attribute__((unused)) TSLexer* lexer,
-    __attribute__((unused)) const bool* valid_symbols) {
+    __attribute__((unused)) TSLexer* lexer, const bool* valid_symbols) {
   // skip past whitespace if any
   bool whitespace = false;
   while (is_whitespace(lexer->lookahead)) {
     whitespace = true;
     lexer->advance(lexer, /* skip= */ true);
   }
 
-  // if any other symbol than *, fallback to treesitter internal lexer
-  if (lexer->lookahead != '*') {
-    return false;
-  }
+  if (lexer->lookahead == '*' &&
+      (valid_symbols[BINARY_STAR] || valid_symbols[POSTFIX_STAR])) {
+    // move to past the *, add * to current token
+    lexer->advance(lexer, /* skip= */ false);
 
-  // move to past the *, add * to current token
-  lexer->advance(lexer, /* skip= */ false);
-
-  // https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/lexical_conventions/symbolic_tokens.md
-  if (is_whitespace(lexer->lookahead) && whitespace) {
-    // foo * bar
-    lexer->result_symbol = BINARY_STAR;
-  } else if (!whitespace &&
-             token_allowed_after_binary_operator(lexer->lookahead)) {
-    // foo*bar or foo*(bar)
-    lexer->result_symbol = BINARY_STAR;
-  } else {
-    // foo*
-    lexer->result_symbol = POSTFIX_STAR;
+    // https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/lexical_conventions/symbolic_tokens.md
+    if (is_whitespace(lexer->lookahead) && whitespace) {
+      // foo * bar
+      lexer->result_symbol = BINARY_STAR;
+    } else if (!whitespace &&
+               token_allowed_after_binary_operator(lexer->lookahead)) {
+      // foo*bar or foo*(bar)
+      lexer->result_symbol = BINARY_STAR;
+    } else {
+      // foo*
+      lexer->result_symbol = POSTFIX_STAR;
+    }
+    return true;
+  } else if ((lexer->lookahead == '#' || lexer->lookahead == '\'' ||
+              lexer->lookahead == '\"') &&
+             valid_symbols[STRING]) {
+    lexer->result_symbol = STRING;
+    int hash_count = 0;
+    while (lexer->lookahead == '#') {
+      advance(lexer);
+      hash_count++;
+    }
+    if (lexer->lookahead == '\'') {
+      if (!eat(lexer, "'''")) {
```

Comment:
Your calls to `eat` could be `eat_count(lexer, '\'', 3)` to only have one function, though maybe that is less clear. Advantage of only using `eat_count` is it could return a number (either the number matched or unmatched) instead of a bool, which would mean information wasn't lost.